lci_strategic_plugin/src/lci_strategic_plugin.cpp

```diff
@@ -753,191 +755,94 @@ void LCIStrategicPlugin::planWhenAPPROACHING(const cav_srvs::PlanManeuversReques
   {
-    ROS_DEBUG_STREAM("Earliest entry time... has gone past the cashed entering time. nearest_green_entry_time_cached_ (which can also be TDB):" << std::to_string(nearest_green_entry_time_cached_.get().toSec()) << ", and earliest_entry_time: " << std::to_string(earliest_entry_time.toSec()));
-  }
+    if (in_tbd) // Given ET is in TBD, but vehicle is too close to intersection
+    {
+      ROS_DEBUG_STREAM("ET is still in TBD despite the vehicle being in desired distance to start stopping. Trying to handle this edge case gracefully...");
+    }
 
-  ROS_DEBUG_STREAM("Final nearest_green_entry_time: " << std::to_string(nearest_green_entry_time.toSec()));
+    double stopping_time = current_state.speed / max_comfort_decel_norm_;
 
-  auto et_state = traffic_light->predictState(lanelet::time::timeFromSec(nearest_green_entry_time.toSec()));
-  ROS_DEBUG_STREAM("Signal at ET: " << et_state.get().second);
+    ros::Time stopping_arrival_time =
+          current_state.stamp + ros::Duration(stopping_time);
 
-  double remaining_time = nearest_green_entry_time.toSec() - current_state.stamp.toSec();
-  double remaining_time_earliest_entry = earliest_entry_time.toSec() - current_state.stamp.toSec();
-  scheduled_entry_time_ = remaining_time; // performance metric
-  earliest_entry_time_ = remaining_time_earliest_entry; // performance metric
-  
-  // CASE SELECTION START
-  auto boundary_distances = get_delta_x(current_state_speed, intersection_speed_.get(), speed_limit, config_.algo_minimum_speed, max_comfort_accel_, max_comfort_decel_);
-  print_boundary_distances(boundary_distances); //debug
+    ROS_DEBUG_STREAM("stopping_arrival_time: " << std::to_string(stopping_arrival_time.toSec()));
 
-  auto boundary_traj_params = get_boundary_traj_params(req.header.stamp.toSec(), current_state_speed, intersection_speed_.get(), speed_limit, config_.algo_minimum_speed, max_comfort_accel_, max_comfort_decel_, current_state.downtrack, traffic_light_down_track, distance_remaining_to_traffic_light, boundary_distances);
+    auto stopping_arrival_state_optional = traffic_light->predictState(lanelet::time::timeFromSec(stopping_arrival_time.toSec()));
 
-  TrajectoryParams ts_params = get_ts_case(req.header.stamp.toSec(), nearest_green_entry_time.toSec(), current_state_speed, intersection_speed_.get(), speed_limit, config_.algo_minimum_speed, max_comfort_accel_, max_comfort_decel_, current_state.downtrack, traffic_light_down_track, distance_remaining_to_traffic_light, boundary_distances, boundary_traj_params);
-  print_params(ts_params);
+    if (!validLightState(stopping_arrival_state_optional, stopping_arrival_time))
+    {
+      ROS_ERROR_STREAM("Unable to resolve give signal for stopping_arrival_state_optional: " << std::to_string(stopping_arrival_time.toSec()));
+      return;
+    }
 
-  ROS_DEBUG_STREAM("SPEED PROFILE CASE:" << ts_params.case_num);
+    if (stopping_arrival_state_optional.get().second == lanelet::CarmaTrafficSignalState::STOP_AND_REMAIN)
```

Comment:
Did we decide to get rid of your earlier fix and replace it with red light safeguard? Given this new changes, we can force the vehicle to stop if the vehicle is within the stopping distance but its ET is within a green. I thought at the minimum we need your previous changes and the plan is to add the red light safeguard on top of it. Please clarify what is the plan? 
If this is still confusing, I would say we should have a quick meeting with the team about this.

Reply:
Also, here you probably want to add yellow light state as well.

Reply:
Is that the case here? It looks like it's projecting the estimated arrival time at the stop bar forward into the known signal timing data, so that if we're expected to arrive on a green we can continue to approach the stop bar even if it's currently red.

Reply:
The projection definitely helps reducing the chance of the edge case I mentioned, but it is still possible.

Another edge case to think about is when a vehicle is trying to enter the intersection box at the end of a green. Imagine a case where the vehicle ET is around the end of the green and the vehicle can technically enter during the current green. With this check alone on top of everything, the estimated stopping arrival time could be within red while the vehicle ET is within the current green and the vehicle can speed up and enter the intersection box.

Sorry if it feels like I am repeating stuff here again, but I guess I am finding new stuff as I think more about the safeguard.